crates/rome_js_parser/src/syntax/jsx/mod.rs

```
@@ -103,8 +107,12 @@ fn parse_any_jsx_tag(p: &mut Parser, in_expression: bool) -> ParsedSyntax {
 
             parse_jsx_children(p);
 
-            expect_closing_element(p, in_expression, name, opening_range);
-            Present(element.complete(p, JSX_ELEMENT))
+            if expect_closing_element(p, in_expression, name, opening_range).is_none() {
+                element.abandon(p);
+                Absent
+            } else {
+                Present(element.complete(p, JSX_ELEMENT))
+            }
```

This is dangerous. It will progress the parser but return `Absent`. The processed tokens now get lost if the caller isn't using a checkpoint. Yes, you added a checkpoint inside of `parse_expression` but this function is also called when parsing the children of an element, in which case the parser silently skips over content.